Sources/Networking/Core/APIManager.swift

```diff
@@ -50,17 +50,119 @@ open class APIManager: APIManaging {
+            let downloadObserver = DownloadObserver()
+            downloadTask.delegate = downloadObserver
+            
+            /// downloadTask must be initiated by resume() before we try to await a response from downloadObserver, because it gets the response from URLSessionDownloadDelegate methods
+            downloadTask.resume()
+            
+            let response = try await downloadObserver.response()
+            
+            do {
+                /// process response. Since we are not returning the response, we don't need to keep reference to it.
+                _ = try await responseProcessors.process((Data(), response), with: request, for: endpointRequest)
+            } catch {
+                /// in case response processing fails we need to manually cancel also the downloadTask otherwise it would just continue running.
+                downloadTask.cancel()
+                throw error
+            }
+            
+            /// reset retry count
+            await retryCounter.reset(for: endpointRequest.id)
+            
+            /// create download AsyncStream
+            return downloadStream(for: endpointRequest, downloadTask: downloadTask, downloadObserver: downloadObserver)
+        } catch {
+            do {
+                /// If retry fails (retryCount is 0 or Task.sleep thrown), catch the error and process it with `ErrorProcessing` plugins.
+                try await sleepIfRetry(for: error, endpointRequest: endpointRequest, retryConfiguration: retryConfiguration)
+            } catch {
+                /// error processing
+                throw await errorProcessors.process(error, for: endpointRequest)
+            }
+            
+            return try await downloadStream(endpointRequest, resumableData: resumableData, retryConfiguration: retryConfiguration)
+        }
+    }
+    
+    func downloadStream(
+        for endpointRequest: EndpointRequest,
+        downloadTask: URLSessionDownloadTask,
+        downloadObserver: DownloadObserver
+    ) -> AsyncThrowingStream<DownloadState, Error>  {
+        return AsyncThrowingStream { continuation in
+            downloadObserver.progressHandler = { (downloaded, total) in
+                continuation.yield(.progress(downloadedBytes: Double(downloaded), totalBytes: Double(total)))
+            }
+
+            downloadObserver.completionHandler = { data in
+                continuation.yield(.completed(data))
+                continuation.finish()
+            }
+            
+            downloadObserver.errorHandler = { [errorProcessors] error in
+                if let resumableData = (error as? URLError)?.userInfo[NSURLSessionDownloadTaskResumeData] as? Data {
+                    continuation.yield(.terminated(resumableData: resumableData))
+                }
+                
+                Task {
+                    /// error processing
+                    let error = await errorProcessors.process(error, for: endpointRequest)
```

Comment:
I've added a retry logic in case something fails during the request creating or response processing, however I am not convinced that we should also add retry logic here in case the download fails while already in progress. Handling temporary loss of network while on poor connection should be handled automatically by URLSession based on it's configuration by setting an appropriate timeout interval for example. For downloads of larger files a background URLSessionConfiguration should be used https://developer.apple.com/documentation/foundation/url_loading_system/downloading_files_in_the_background
What do you think @cejanen?